File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n==1:
            return 0
        
        buyPrice = prices[0]
        profit=0
        for i in range(1, n):
            
            if prices[i]<buyPrice:
                buyPrice = prices[i]
                continue
            
            if prices[i]> buyPrice and prices[i] -buyPrice > profit:
                profit= prices[i]-buyPrice
                    
        return profit
                
        

File: Easy/121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        if (prices.Length ==1){
            return 0;
        }
        
        int maxSoFar = prices[prices.Length-1];
        
        int[] arr = new int[prices.Length];
        arr[arr.Length-1] = 0;
        for (int i = prices.Length-2; i>=0; i--){
            if (prices[i] > maxSoFar) {
                arr[i] = 0; maxSoFar = prices[i];
            } else{
                arr[i] = maxSoFar - prices[i];
            }
        }
        maxSoFar =0;
        foreach (int a in arr) {
            maxSoFar = Math.Max(maxSoFar, a);
        }
        return maxSoFar;
    }
}